### PR TITLE
Save/restore quiz progress

### PR DESCRIPTION
**Pull request by kdeloach**
Sticks the currently selected quiz answers into localStorage so that quizzes
can be resumed later.

### CODE REVIEW

File: src/nyc_trees/js/src/quizProgress.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        var values = deserialize(storage[key] || '');
+
+        // Restore radio button selections.
+        $.each(values, function(fieldName, value) {
+            // Assumes that only checked values are serialized.
+            var radio = $('input[name="' + fieldName + '"][value="' + value + '"]');
+            radio.prop('checked', true);
+        });
+
+        // Return currently active question.
```
Comment (kdeloach) [inline src/nyc_trees/js/src/quizProgress.js]: TODO: Move up to function def.

File: src/nyc_trees/js/src/quizProgress.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        return activeQuestionNumber;
+    }
+
+    // Delete serialized data from storage.
+    function clear() {
+        delete storage[key];
+    }
+
+    // Return quiz selections serialized into a string.
+    function serialize() {
```
Comment (kdeloach) [inline src/nyc_trees/js/src/quizProgress.js]: TODO: Accept selections as argument and add unit test.

File: src/nyc_trees/js/src/quizProgress.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+function QuizProgress(slug) {
+    var key = 'quiz-progress-' + slug;
+
+    function save() {
+        storage[key] = serialize();
+    }
+
+    function restore() {
+        var values = deserialize(storage[key] || '');
```
Comment (kdeloach) [inline src/nyc_trees/js/src/quizProgress.js]: TODO: Move this fallback to def.

**Comment by kdeloach** [comment]
@azavea-bot rebuild

**Comment by jwalgran** [comment]
Checking this out now

**Comment by jwalgran** [comment]
+1 this works well

This set of actions produces a result worth talking about:

1. Answer Q1, click next
1. Answer Q2, click next
1. Answer Q3
1. Click back twice to return to the first question (I'm checking my work)
1. Refresh the page

The progress system jumps you to Q3. This is a little jarring. Serializing "current question" as a separate value seems like a good idea. Is there hidden complexity to adding this feature?

**Comment by kdeloach** [comment]
There's no hidden complexity, I'll add it in.

**Comment by kdeloach** [comment]
OK I think it works a lot better now that the current step is serialized. I was also able to clean up the code a bit to make things simpler. May want to take another look before it gets merged in.

**Comment by jwalgran** [comment]
Nice updates. Still works.

**Comment by maurizi** [comment]
Looks good to me

**Pull request merged**